logpush-to-bigquery/index.js

```diff
@@ -26,17 +22,75 @@ async function gcsbq (file, context) {
     ignoreUnknownValues: true
   }
 
-  const addToTable = async (tableId) => {
+  const addToTable = async tableId => {
     const dataset = await bigquery.dataset(datasetId).get({ autoCreate: true })
     const table = await dataset[0].table(tableId).get({ autoCreate: true })
-    return table[0].load(filename, metadata)
+    return table[0].createLoadJob(files, metadata)
   }
 
   try {
-    await addToTable(tableId)
+    return addToTable(tableId)
   } catch (e) {
     console.log(e)
   }
 }
 
-module.exports.gcsbq = gcsbq
+async function writeLog (logData) {
+  logData = logData.join(',\n')
+
+  const metadata = {
+    resource: { type: 'global' },
+    severity: 'INFO'
+  }
+
+  const entry = log.entry(metadata, logData)
+  await log.write(entry)
+  console.log(`Logged: ${logData}`)
+}
+
+module.exports.runLoadJob = async function (message, context) {
+  if (!context) {
+    context = {}
+    context.timestamp = new Date().toISOString()
+  }
+  context.timestamp = DateTime.fromISO(context.timestamp)
+
+  const loadJobDeadline = context.timestamp
+    .setZone('GMT')
+    .minus({ minutes: 3 })
+    .startOf('minute')
+
+  const [deadlineDate, deadlineTime] = [
+    loadJobDeadline.toFormat('yyyyMMdd'),
+    loadJobDeadline.toFormat(`yyyyMMdd'T'hhmm`)
+  ]
+
+  let arrayOfLogpushLogs = []
+  let stackdriverEntry = []
+
+  try {
+    bucket
+      .getFilesStream({
+        autoPaginate: false,
+        maxResults: 5000,
+        prefix: `${process.env.DIRECTORY}${deadlineDate}/${deadlineTime}`
+      })
+      .on('error', console.error)
+      .on('data', function (file) {
+        let logEndTime = file.name.split('_').slice(1, -1)
+        logEndTime = DateTime.fromISO(logEndTime, { setZone: true })
```

Comment:
Where is logEndTime used?

Reply:
I inadvertently removed an `if` condition here. It should have been:
```js
logEndTime = DateTime.fromISO(logEndTime, { setZone: true })
if (deadlineTime.hasSame(logEndTime, 'minute')) {
  arrayOfLogpushLogs.push(bucket.file(file.name))
}
```
However this condition was a temporary patch to prototype this function anyways. `logEndTime` needs to be inserted as part of the query passed to `getFiles()` or `getFileStream()`. As you mention above, we need a solution that allows a wildcard or pattern match to make this work at scale.

Reply:
Ah ha! You can do it with the `gsutil` CLI tool:
```sh
gsutil ls -r gs://burritobotschemanew/logs/20210419/20210419T**Z_20210419T000**.log.gz
```
This is probably wrapping an API call -- though it may be possible to spawn a child process within the Node runtime itself and invoke `gsutil`. Will investigate